### DISCUSSION TITLE
unhandled exception in mqtt.py , puback method

### DISCUSSION THREAD

**Discussion by RoccoCloudCycle** [awslabs/aws-crt-python#572] [Q&A]
Hello,
in mqtt.py, if Connetcion.publish() is called and then the associated future is cancelled, the method will still process the response which will raise an exception:
`def puback(packet_id, error_code):
            if error_code != 0:
                future.set_exception(awscrt.exceptions.from_code(error_code))
            else:
                future.set_result(dict(packet_id=packet_id)) `

 
`    def set_result(self, result):
        with self._condition:
            if self._state in {CANCELLED, CANCELLED_AND_NOTIFIED, FINISHED}:
                raise InvalidStateError('{}: {!r}'.format(self._state, self))
             .....`

at runtime I get the following message:
`
Exception ignored in: <class 'concurrent.futures._base.InvalidStateError'>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/awscrt/mqtt.py", line 810, in puback
    future.set_result(dict(packet_id=packet_id))
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 544, in set_result
    raise InvalidStateError('{}: {!r}'.format(self._state, self))
concurrent.futures._base.InvalidStateError: CANCELLED: <Future at 0xffff74199210 state=cancelled>`

why the exception is not handled? should the error_code be set when the futures is cancelled?

Regards,
Rocco

**Comment by bretambrose** [comment]
I'm of two minds here.

* On one hand, using a valid future API (cancel) shouldn't cause an exception later on.
* On the other hand, cancel does nothing on these futures (and likely never will have any effect), so from a certain standpoint, this is a trivially resolvable issue: don't call cancel on them.  That being said, perhaps you might be in a situation where the future is actually  a piece of a larger aggregate future that is cancellable and so you're not able to just "don't call cancel."

We could swallow these exceptions internally I suppose.

**Reply by RoccoCloudCycle** [reply]
hello, sorry for the delay in responding.
maybe a little bit of context could help: our IoT devices are mounted on trucks and connected via modem to AWS. this means we have a lot of disconnection are reconnection caused by the devices moving around and situations of low bandwidth. what we are experiencing is that sometimes the futures are hanging indefinitely, maybe because the response from the broker has been lost due to modem disconnection. so what we are trying to achieve is to cancel these pending futures and resend them again
